Fixes #1169: AES-GCM implementation (#1291)
opentofu/opentofu: website/docs/language/state/encryption.mdx

### What does encryption protect against?

When you enable encryption, OpenTofu will encrypt state data *at rest*. If an attacker were to gain access to your state file, they should not be able to read it and use the sensitive values (e.g. access keys) contained in the state file.

However, encryption does not protect against data loss (your state file getting damaged) and it also does not protect against replay attack (an attacker using an older state or plan file and tricking you into running it). Additionally, OpenTofu does not and cannot protect the sensitive values in the state file from the person running the `tofu` command.

### What precautions do I need to take?

When you enable encryption, consider who needs access to your state file directly. If you have more than a very small number of people with access needs, you may want to consider running your production `plan` and `apply` runs from a continuous integration system to protect both the encryption key and the sensitive values in your state.

> You will also need to decide what kind of key you would like to use. You can either opt for a static passphrase which you manually rotate regularly, or you can choose a key management system. If feasible, we recommend using a key management system that can rotate keys automatically to give you the maximum amount of protection.

Reviewer: I don't think there's any proper reason to regularly rotate passphrases, is there?

With key derivation we'll get a new encryption key every time anyway.

Reviewer 2: See https://github.com/opentofu/opentofu/pull/1291#discussion_r1512927813 -- there's the same reason to rotate passwords as for phrases, unless additional data is mixed in to the password to make it unique.

Likely there isn't, otherwise you could use a KBKDF in place of a PBKDF and get the same properties, IMHO.

> If you decide for a key management system, please make sure to configure regular key rotation. This is especially important if the chosen encryption method can exhaust the key. (This is not a problem with passphrase-based static keys.) You can find more information in the [encryption methods](#encryption-methods) section below.

Reviewer 3: ```suggestion
If you opt for a key management system, it is imperative to configure automatic key rotation. This is particularly crucial if the encryption algorithm you choose has the potential to reach a point of 'key saturation', where the maximum safe usage limit of the key is approached. You can find more information about this in the [encryption methods](#encryption-methods) section below.
```

> If you decide for a key management system, please make sure to configure regular key rotation. This is especially important if the chosen encryption method can exhaust the key. (This is not a problem with passphrase-based static keys.) You can find more information in the [encryption methods](#encryption-methods) section below.

Author: ```suggestion
### What precautions do I need to take?

When you enable encryption, consider who needs access to your state file directly. If you have more than a very small number of people with access needs, you may want to consider running your production `plan` and `apply` runs from a continuous integration system to protect both the encryption key and the sensitive values in your state.

You will also need to decide what kind of key you would like to use based on your security requirements. You can either opt for a static passphrase or you can choose a key management system. If you opt for a key management system, it is imperative to configure automatic key rotation for some encryption methods. This is particularly crucial if the encryption algorithm you choose has the potential to reach a point of 'key saturation', where the maximum safe usage limit of the key is approached, such as AES-GCM. You can find more information about this in the [encryption methods](#encryption-methods) section below.

Finally, before enable encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. You should also have backups of your keys. Once you enable encryption, OpenTofu cannot read your state file without the correct key.
```

> 

Reviewer 4: I'd suggest adding paragraph about backing up keys and the configuration used for encryption. If either of those are lost, the state data may be unrecoverable. Maybe something like:

"Ensure you will be able to recover both the keys (passphrase, kms data, ...) and the relevant configuration (code, environment variable values) used to encrypt your state."

> Finally, before enable encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. Once you enable encryption, OpenTofu cannot read your state file without the correct key.

Reviewer: before enable => before enabling

> Finally, before enable encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. Once you enable encryption, OpenTofu cannot read your state file without the correct key.

Author: ```suggestion
### What precautions do I need to take?

When you enable encryption, consider who needs access to your state file directly. If you have more than a very small number of people with access needs, you may want to consider running your production `plan` and `apply` runs from a continuous integration system to protect both the encryption key and the sensitive values in your state.

You will also need to decide what kind of key you would like to use based on your security requirements. You can either opt for a static passphrase or you can choose a key management system. If you opt for a key management system, it is imperative to configure automatic key rotation for some encryption methods. This is particularly crucial if the encryption algorithm you choose has the potential to reach a point of 'key saturation', where the maximum safe usage limit of the key is approached, such as AES-GCM. You can find more information about this in the [encryption methods](#encryption-methods) section below.

Finally, before enabling encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. You should also have backups of your keys. Once you enable encryption, OpenTofu cannot read your state file without the correct key.

```

> Finally, before enable encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. Once you enable encryption, OpenTofu cannot read your state file without the correct key.

Author: ```suggestion
### What precautions do I need to take?

When you enable encryption, consider who needs access to your state file directly. If you have more than a very small number of people with access needs, you may want to consider running your production `plan` and `apply` runs from a continuous integration system to protect both the encryption key and the sensitive values in your state.

You will also need to decide what kind of key you would like to use based on your security requirements. You can either opt for a static passphrase or you can choose a key management system. If you opt for a key management system, it is imperative to configure automatic key rotation for some encryption methods. This is particularly crucial if the encryption algorithm you choose has the potential to reach a point of 'key saturation', where the maximum safe usage limit of the key is approached, such as AES-GCM. You can find more information about this in the [encryption methods](#encryption-methods) section below.

Finally, before enabling encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. You should also have backups of your keys. Once you enable encryption, OpenTofu cannot read your state file without the correct key.
```

## Configuration

You can configure encryption in OpenTofu either by specifying the configuration in the OpenTofu code, or using the `TF_ENCRYPTION` environment variable. Both solutions are equivalent and if you use both, OpenTofu will merge the two configurations, overriding any code-based settings with the environment ones.

The basic configuration structure looks as follows:

<Tabs>
    <TabItem value="code" label="Code" default>
        <CodeBlock language={"hcl"}>{ConfigurationTF}</CodeBlock>
    </TabItem>
    <TabItem value="env-sh" label="Environment (Linux/UNIX shell)">
        <CodeBlock language={"shell"}>{ConfigurationSH}</CodeBlock>
    </TabItem>
    <TabItem value="env-ps1" label="Environment (Powershell)">
        <CodeBlock language={"powershell"}>{ConfigurationPS1}</CodeBlock>
    </TabItem>
</Tabs>

:::warning

> Once your data is encrypted, do not rename key providers and methods in your configuration! The encrypted data contains metadata related to their specific names. Instead, use a [fallback block](#key-and-method-rollover).

Reviewer 3: ```suggestion
Once your data is encrypted, you should not rename key providers and methods in your configuration! The encrypted data stored in the backend contains metadata related to their specific names. Instead you should use a [fallback block](#key-and-method-rollover) to handled changes to key providers.

```

:::

:::tip

You can use the [JSON configuration syntax](/docs/language/syntax/json/) instead of HCL for encryption configuration.

:::

:::tip

If you use environment configuration, you can include the following code configuration to prevent unencrypted data from being written in the absence of an environment variable:

<CodeBlock language="hcl">{Enforce}</CodeBlock>

:::

## Key providers

There are currently no supported key providers.

## Methods

### AES-GCM

The only currently supported encryption method is AES-GCM. You can configure it in the following way:

<CodeBlock language="hcl">{AESGCM}</CodeBlock>

:::warning

> When using a key-management system as a key provider, always configure automatic key rotation in conjunction with AES-GCM, otherwise your security may be compromised! (This does not affect passphrase-based local keys.)

Reviewer: Frankly, with our official implementations and no plugins, I don't think this will affect **any** key providers.

Reviewer: Since they will be generating a new data encryption key every time.

Author: ```suggestion
AES-GCM is a secure, industry-standard encryption algorithm, but suffers from "key saturation". In order to configure a secure setup, you should either use a key-derivation key provider (such as PBKDF2) with a long and complex passphrase, or use a key management system that automatically rotates keys regularly. Using short, static keys will degrade your encryption.
```

> When using a key-management system as a key provider, always configure automatic key rotation in conjunction with AES-GCM, otherwise your security may be compromised! (This does not affect passphrase-based local keys.)

Reviewer 3: this is a repeat of L37

:::

## Key and method rollover

In some cases, you may want to change your encryption configuration. This can include renaming a key provider or method, changing a passphrase for a key provider, or switching key-management systems. OpenTofu supports an automatic rollover of your encryption configuration if you provide your old configuration in a `fallback` block:

<CodeBlock language="hcl">{Fallback}</CodeBlock>

> If OpenTofu fails to read your state or plan file with the new method, it automatically tries the fallback method. When OpenTofu saves your state or plan files, it will always use the new method.

Reviewer 3: Maybe we should embolden read and saves here to highlight the difference in actions being the driving factor behind the differences.

## Remote state data sources

> For projects using the `terraform_remote_state` data source, you can also configure encryption. This can be the same encryption setup as your main configuration, but you can also define a separate set of keys and methods. The configuration syntax looks as follows:

Reviewer 3: ```suggestion
You can also configure an encryption setup for projects using the `terraform_remote_state` data source. This can be the same encryption setup as your main configuration, but you can also define a separate set of keys and methods. The configuration syntax looks as follows:

```

:::

:::tip

If you use environment configuration, you can include the following code configuration to prevent unencrypted data from being written in the absence of an environment variable:

<CodeBlock language="hcl">{Enforce}</CodeBlock>

:::

## Key providers

There are currently no supported key providers.

## Methods

### AES-GCM

The only currently supported encryption method is AES-GCM. You can configure it in the following way:

<CodeBlock language="hcl">{AESGCM}</CodeBlock>

:::warning

When using a key-management system as a key provider, always configure automatic key rotation in conjunction with AES-GCM, otherwise your security may be compromised! (This does not affect passphrase-based local keys.)

:::

## Key and method rollover

In some cases, you may want to change your encryption configuration. This can include renaming a key provider or method, changing a passphrase for a key provider, or switching key-management systems. OpenTofu supports an automatic rollover of your encryption configuration if you provide your old configuration in a `fallback` block:

<CodeBlock language="hcl">{Fallback}</CodeBlock>

> If OpenTofu fails to read your state or plan file with the new method, it automatically tries the fallback method. When OpenTofu saves your state or plan files, it will always use the new method.

Reviewer 3: ```suggestion
If OpenTofu fails to **read** your state or plan file with the new method, it will automatically try the fallback method. When OpenTofu **saves** your state or plan file, it will always use the new method and not the fallback.

```

OpenTofu supports encrypting state and plan files at rest, both for local storage and when using a backend. In addition, you can also use encryption with the `terraform_remote_state` data source. This page explains how to set up encryption and what encryption method is suitable for which use case.

## General guidance and pitfalls (please read)

When you enable encryption, your state and plan files become unrecoverable without the appropriate encryption key. Please make sure you read this section carefully before enabling encryption.

### What does encryption protect against?

When you enable encryption, OpenTofu will encrypt state data *at rest*. If an attacker were to gain access to your state file, they should not be able to read it and use the sensitive values (e.g. access keys) contained in the state file.

However, encryption does not protect against data loss (your state file getting damaged) and it also does not protect against replay attack (an attacker using an older state or plan file and tricking you into running it). Additionally, OpenTofu does not and cannot protect the sensitive values in the state file from the person running the `tofu` command.

> ### What precautions do I need to take?
> 
> When you enable encryption, consider who needs access to your state file directly. If you have more than a very small number of people with access needs, you may want to consider running your production `plan` and `apply` runs from a continuous integration system to protect both the encryption key and the sensitive values in your state.
> 
> You will also need to decide what kind of key you would like to use. You can either opt for a static passphrase which you manually rotate regularly, or you can choose a key management system. If feasible, we recommend using a key management system that can rotate keys automatically to give you the maximum amount of protection.
> 
> If you opt for a key management system, it is imperative to configure automatic key rotation. This is particularly crucial if the encryption algorithm you choose has the potential to reach a point of 'key saturation', where the maximum safe usage limit of the key is approached. You can find more information about this in the [encryption methods](#encryption-methods) section below.
> 
> Finally, before enable encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. Once you enable encryption, OpenTofu cannot read your state file without the correct key.

Author: ```suggestion
### What precautions do I need to take?

When you enable encryption, consider who needs access to your state file directly. If you have more than a very small number of people with access needs, you may want to consider running your production `plan` and `apply` runs from a continuous integration system to protect both the encryption key and the sensitive values in your state.

You will also need to decide what kind of key you would like to use based on your security requirements. You can either opt for a static passphrase or you can choose a key management system. If you opt for a key management system, it is imperative to configure automatic key rotation for some encryption methods. This is particularly crucial if the encryption algorithm you choose has the potential to reach a point of 'key saturation', where the maximum safe usage limit of the key is approached, such as AES-GCM. You can find more information about this in the [encryption methods](#encryption-methods) section below.

Finally, before enabling encryption, please exercise your disaster recovery plan and make a temporary backup of your unencrypted state file. You should also have backups of your keys. Once you enable encryption, OpenTofu cannot read your state file without the correct key.
```
